### PR TITLE
Introduce span and trace ID logs injection

### PR DESCRIPTION
**Pull request by owais**
Link individual log entries with trace IDs and span IDs associated with corresponding events. The SignalFx Python instrumentation patches `logging.Logger.makeRecord` method to automatically inject trace context into all `LogRecord` objects. It also sets the default logging format to include the span and trace IDs. The default format looks like the following:

```
%(asctime)s %(levelname)s [%(name)s] [%(filename)s:%(lineno)d] [signalfx.trace_id=%(sfx_trace_id)s signalfx.span_id=%(sfx_span_id)s] - %(message)s
```

Log injection is not enabled by default and can be enabled by setting `SIGNALFX_LOGS_INJECTION` environment variable to `true`.

If you are using a custom logging hanlder or format, you'll need to manually include span ID and trace ID placeholders in the logging format.

### CODE REVIEW

**Comment by rmfitzpatrick** [comment]
Would you be able to treat this as a first class instrumentation as opposed to a flagged feature and migrate to `libraries/logging_` to prevent the collision?

File: README.md [hunk 1]
```diff
@@ -262,3 +262,15 @@ deploying in a test environment.
...
+
+Link individual log entries with trace IDs and span IDs associated with corresponding events. The SignalFx Python instrumentation patches `logging.Logger.makeRecord` method to automatically inject trace context into all `LogRecord` objects. It also sets the default logging format to include the span and trace IDs. The default format looks like the following:
+
+```
+%(asctime)s %(levelname)s [%(name)s] [%(filename)s:%(lineno)d] [signalfx.trace_id=%(sfx_trace_id)s signalfx.span_id=%(sfx_span_id)s] - %(message)s
+```
+
+Log injection is not enabled by default and can be enabled by setting `SIGNALFX_LOGS_INJECTION` environment variable to `true`.
+
+If you are using a custom logging hanlder or format, you'll need to manually include span ID and trace ID placeholders in the logging format.
```
Comment (rmfitzpatrick) [inline README.md:276]: nit: appears that LogRecord format strings opt for lower or camel over snake case, so how about `signalfxtraceid` or `signalfxTraceId` as the attributes?  The java-tracing MDC instrumentation is using `signalfx.trace_id` and `signalfx.span_id` as the keys like your templated prefixes so full name seems closer to establishing a convention.

File: signalfx_tracing/instrumentation.py [hunk 2]
```diff
@@ -19,6 +20,10 @@ def _tracing_enabled():
     return is_truthy(os.environ.get('SIGNALFX_TRACING_ENABLED', True))
 
 
+def _logs_injection_enabled():
+    return is_truthy(os.environ.get('SIGNALFX_LOGS_INJECTION', False))
```
Comment (rmfitzpatrick) [inline signalfx_tracing/instrumentation.py:24]: To clarify earlier request, I think this env var should still be supported.

File: signalfx_tracing/logging.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    span = opentracing.tracer.active_span
+    if span is not None:
+        span_id = span.span_id
+        trace_id = span.trace_id
+    setattr(rv, 'sfx_trace_id', span_id)
+    setattr(rv, 'sfx_span_id', trace_id)
+    return rv
+
+
+def enable_logs_injection(tracer=None):
```
Comment (rmfitzpatrick) [inline signalfx_tracing/logging.py:27]: How do you feel about splitting this up into two actions.  One would be to treat the "proper" patch to add the id attributes to the log record for custom format support, and the second would be to use the `SIGNALFX_LOGS_INJECTION` to set the root logger's format string?

**Comment by owais** [comment]
> Would you be able to treat this as a first class instrumentation as opposed to a flagged feature and migrate to libraries/logging_ to prevent the collision?

That was my first implementation as well  but then I thought instrumentation would make sense if we were trying to track usage of the logger library and generate spans from any operations carried out using it. Since we are not "instrumenting" it to generate spans, I wasn't sure if adding it as an instrumentation lib was a good idea. The use case seemed to be distinct and specialized enough to warrant it's own "place" in the code-base. May be we can add concept of propagators to this repo and make this the first supported propagator but then again, I don't see too much value in doing that until we have at least two "propagators" in the repo.

**Comment by owais** [comment]
@rmfitzpatrick  ^

**Comment by rmfitzpatrick** [comment]
@owais, though the behavior is different from other span creating instrumentations, it's still a wrapt-utilizing instrumentation with a specific target and potentially configurable behavior, so I'd prefer the existing machinery be utilized.  I agree and don't think we need to create any special propagation helper for it.  The relocated module would also resolve the apparent collision in python 2.7.

File: tests/unit/test_instrument.py [hunk 4]
```diff
@@ -126,7 +126,7 @@ def test_instrument_with_true_and_env_var_false_doesnt_instrument_specified_libr
             os.environ.pop(env_var)
 
     def test_auto_instrument_instruments_all_available_libraries(self):
-        modules = [(utils.get_module(l), l) for l in expected_traceable_libraries]
+        modules = [(utils.get_module(lib), lib) for lib in expected_traceable_libraries]
```
Comment (rmfitzpatrick) [inline tests/unit/test_instrument.py:129]: thanks

File: signalfx_tracing/logging.py (hunk 3 above)
Comment (owais) [reply signalfx_tracing/logging.py]: I don't see much benefit in doing that especially now that we moved the patch into an "instrumentation" package. Setting the format depends on the patch. If the format is set without applying the patch, logging will raise exceptions.

**Comment by owais** [comment]
@rmfitzpatrick implemented as an instrumentation package.

File: signalfx_tracing/libraries/logging_/instrument.py [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+    logging.basicConfig(level=level, format=fmt)
+
+    utils.mark_instrumented(logging)
+
+
+def uninstrument():
+    logging = utils.get_module('logging')
+    if not utils.is_instrumented(logging):
+        return
+
```
Comment (rmfitzpatrick) [inline signalfx_tracing/libraries/logging_/instrument.py:56]: Mind calling this? https://github.com/signalfx/signalfx-python-tracing/blob/master/signalfx_tracing/utils.py#L69

File: signalfx_tracing/libraries/logging_/instrument.py [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+from signalfx_tracing import utils
+from signalfx_tracing.constants import logging_format
+
+
+config = utils.Config()
+
+
+def _logs_injection_enabled():
+    return utils.is_truthy(os.environ.get('SIGNALFX_LOGS_INJECTION', False))
```
Comment (rmfitzpatrick) [inline signalfx_tracing/libraries/logging_/instrument.py:13]: Mind providing a `logs_injection` config parameter w/ this as a default?

File: signalfx_tracing/libraries/logging_/instrument.py [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+
+
+def makeRecordPatched(makeRecord, instance, args, kwargs):
+    rv = makeRecord(*args, **kwargs)
+    span_id = ''
+    trace_id = ''
+    span = opentracing.tracer.active_span
+    if span is not None:
+        span_id = hex(span.span_id)
+        trace_id = hex(span.trace_id)
```
Comment (rmfitzpatrick) [inline signalfx_tracing/libraries/logging_/instrument.py:23]: for better experience these should be left padded to allow c/p in the traces endpoint and ui.

File: signalfx_tracing/libraries/logging_/instrument.py [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    setattr(rv, 'sfx_trace_id', span_id)
+    setattr(rv, 'sfx_span_id', trace_id)
```
Comment (rmfitzpatrick) [inline signalfx_tracing/libraries/logging_/instrument.py:25]: https://github.com/signalfx/signalfx-python-tracing/pull/43/files#r450317432

File: signalfx_tracing/libraries/logging_/instrument.py [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+        return
+
+    if not _logs_injection_enabled():
+        return
+
+    wrap_function_wrapper('logging', 'Logger.makeRecord', makeRecordPatched)
+    level = logging.INFO
+    if utils.is_truthy(os.environ.get('SIGNALFX_TRACING_DEBUG', False)):
+        level = logging.DEBUG
+    fmt = os.environ.get('SIGNALFX_LOGGING_FORMAT', logging_format)
```
Comment (rmfitzpatrick) [inline signalfx_tracing/libraries/logging_/instrument.py:46]: Could also be set as a config parameter and documented in an additional logging instrumentation readme

File: noxfile.py [hunk 10]
```diff
@@ -34,7 +34,7 @@ def build(session):
 
 def install_unit_tests(session, *other_packaages):
     build(session)
-    return session.install('attrs==19.1.0', pytest, f'{sdist}[unit_tests]', *other_packaages)
+    return session.install('attrs==19.1.0', 'sfx-jaeger-client', pytest, f'{sdist}[unit_tests]', *other_packaages)
```
Comment (rmfitzpatrick) [inline noxfile.py:37]: Why is this necessary?

File: signalfx_tracing/libraries/logging_/instrument.py (hunk 6 above)
Comment (owais) [reply signalfx_tracing/libraries/logging_/instrument.py:13]: Sure. Can you point me to an example of how config is passed to instrument functions by users? I'm a bit confused. Looks like existing instrumentations use the globally defined `config` variable as the config and I'm not sure how that is supposed to be configured by users.

Comment (owais) [reply signalfx_tracing/libraries/logging_/instrument.py:13]: On second thought, I don't think this should be a config param. If users don't want to enable this, they just won't call the `logging_.instrument()` method in the first place, and if it is automatically instrumented then they can use the env var to disable it (even though it is off by default). 

I guess it depends on how users pass config manually in python code to instrumentions. If there is some mechanism like `auto_instrument(tracer, config={'logging': {...}})` then may be we can add this option but even then I think better way would be to support `auto_instrument(tracer, logging=false)` since all this option does is disable the instrumentation entirely. 

We can support the format string as a config param like you mentioned below though.

Comment (owais) [reply signalfx_tracing/libraries/logging_/instrument.py:13]: OK. Makes sense to add it after our related offline discussion.

File: noxfile.py (hunk 10 above)
Comment (owais) [reply noxfile.py:37]: The test needs span and trace IDs to verify context propagation and using MockTracer does not work as MockSpan does not generate those (AFAICT). Using real tracer enabled testing against span/trace ID.

File: signalfx_tracing/libraries/logging_/instrument.py (hunk 5 above)
Comment (owais) [reply signalfx_tracing/libraries/logging_/instrument.py:56]: That was the intention. Forgot after copy-paste from another instrumentation.

File: README.md (hunk 1 above)
Comment (owais) [reply README.md:276]: I haven't been able to figure out what the convention is but examples across OSS projects show that snake case is at least the norm so I suggest we use `signalfx_span_id`

Comment (owais) [reply README.md:276]: So far only example I've found is structlog which prefers snake case. Haven't seen anyone use camel case.

File: noxfile.py (hunk 10 above)
Comment (rmfitzpatrick) [reply noxfile.py:37]: If the tests now have a jaeger dependency we should probably add it here so the unit_tests extras picks it up automatically: https://github.com/signalfx/signalfx-python-tracing/blob/master/setup.py#L118

File: signalfx_tracing/libraries/logging_/instrument.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+    """
+    Unlike all other instrumentations, this instrumentation does not patch the logging
+    lib to automatically generate spans. Instead it patches the lib to automatically
+    inject trace context into logs.
+    """
+    logging = utils.get_module('logging')
+    if utils.is_instrumented(logging):
+        return
+
+    if not config.enabled:
```
Comment (rmfitzpatrick) [inline signalfx_tracing/libraries/logging_/instrument.py:46]: Sry for being vague, but I think the patching functionality should be the auto-instrumentation element (would be disabled via `SIGNALFX_LOGGING_ENABLED` env var and standard instrument/auto_instrumentation functionality so the ids are available for usage in custom formats by default.  The additional config item and associated `SIGNALFX_LOGS_INJECTION` env var should be for setting the format and log level in the root logger.

File: signalfx_tracing/libraries/logging_/instrument.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
+import os
+from wrapt import wrap_function_wrapper
+import opentracing
+
+from signalfx_tracing import utils
+from signalfx_tracing.constants import logging_format
+
+
+config = utils.Config(
+    enabled=utils.is_truthy(os.environ.get('SIGNALFX_LOGS_INJECTION', False)),
```
Comment (rmfitzpatrick) [inline signalfx_tracing/libraries/logging_/instrument.py:10]: `logs_injection` for clarification?  This being enabled/disabled is now based on the instrument() functionality.

File: signalfx_tracing/libraries/logging_/instrument.py (hunk 11 above)
Comment (owais) [reply signalfx_tracing/libraries/logging_/instrument.py:46]: That would mean `SIGNALFX_LOGS_INJECTION` would always depend on `SIGNALFX_LOGGING_ENABLED`. Enabling the former without the latter would result in crashing applications unless the instrumentation implicitly set `SIGNALFX_LOGGING_ENABLED` to true whenever it saw `SIGNALFX_LOGS_INJECTION` was set to true by the user. Also, is this how other languages support the feature? If not, wouldn't it result in an inconsistent experience for users?

File: signalfx_tracing/libraries/logging_/instrument.py (hunk 12 above)
Comment (owais) [reply signalfx_tracing/libraries/logging_/instrument.py:10]: As it stands now, the only thing the insrumentation package does *is* log injection so I felt naming it log_injection was kinda redundant but I guess depends on outcome of other discussion. https://github.com/signalfx/signalfx-python-tracing/pull/43#discussion_r451617122

File: README.md (hunk 1 above)
Comment (rmfitzpatrick) [reply README.md:276]: I was referring to the stdlib: https://github.com/python/cpython/blob/master/Lib/logging/__init__.py#L517

File: signalfx_tracing/libraries/logging_/instrument.py (hunk 11 above)
Comment (rmfitzpatrick) [reply signalfx_tracing/libraries/logging_/instrument.py:46]: Right, if logs_injection is true and logging_enabled is false then logs_injection would have no effect.  Not sure how it would cause crashes since the functionality occurs in the instrument call that isn't invoked if logging_enabled is false.  For java mdc if logs_injection is false and the id keys are attempted to be used I think that would raise an error so the suggested functionality is a bit better here if I'm following. Node doesn't seem to apply here since the injection occurs manually as you documented or in the supported logger instrumentations themselves.

Comment (owais) [reply signalfx_tracing/libraries/logging_/instrument.py:46]: > Right, if logs_injection is true and logging_enabled is false then logs_injection would have no effect.

I'm not sure if that should be the case. If someone sets logging_enabled to false, it means the instrumentation code will not log any statements but the actual app would still log on it's own and log injection should be available to it. I don't think log injection should depend on logging being enabled for the library. To me they sound like two distinct features. 

On crash: The scenario where this can result in crashes is if logging format is set to render span/trace IDs but makeRecord has not been patched to include those values in LogRecord objects. To me patching makeRecord and setting the logging format to enable injection are part of the same feature. 

We can freely patch `makeRecord` to include any extra values without setting changing the logging format. That would be safe to do as in it won't cause a crash. In any case, setting the log format should only happen automatically when log injection is enabled. We could patch makeRecord unconditionally and set logging format only when log injection is enabled but I don't see much value in that. Anything that can be achieved with that can be also be achieved by enabling log injection and providing custom logging format.

File: signalfx_tracing/libraries/logging_/instrument.py [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+from signalfx_tracing.constants import logging_format
+
+
+config = utils.Config(
+    injection_enabled=utils.is_truthy(os.environ.get('SIGNALFX_LOGS_INJECTION', False)),
+    logging_format=os.environ.get('SIGNALFX_LOGGING_FORMAT', logging_format)
+)
+
+
+def _logs_injection_enabled():
```
Comment (rmfitzpatrick) [inline signalfx_tracing/libraries/logging_/instrument.py:15]: no longer used?
